From ec070f25882f6dadc8fdfcd268dddcd8253cb300 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor+arch@gmail.com>
Date: Thu, 26 Jan 2023 19:56:00 +0100
Subject: [PATCH] sodepends: Handle RPATH better

- The RPATH needs to be normalized, e.g. collapsing ../ parts.
- If the RPATH contains $ORIGIN, we need to substitute it with the path of the linked binary.
---
 Namcap/rules/sodepends.py | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/Namcap/rules/sodepends.py b/Namcap/rules/sodepends.py
index 3b030b9..641c5b9 100644
--- a/Namcap/rules/sodepends.py
+++ b/Namcap/rules/sodepends.py
@@ -150,8 +150,10 @@ class SharedLibsRule(TarballRule):
 				rpaths = list(get_rpaths(f)) + list(get_runpaths(f))
 				f.seek(0)
 				for n in pkg_so_files:
-					if any(n.startswith(rp) for rp in rpaths):
-						rpath_files[os.path.basename(n)] = n
+					for rp in rpaths:
+						rp = os.path.normpath(rp.replace('$ORIGIN', '/' + os.path.dirname(entry.path)))
+						if os.path.dirname(n) == rp:
+							rpath_files[os.path.basename(n)] = n
 			scanlibs(f, entry.name, rpath_files, liblist, libdepends, libprovides)
 			f.close()
 
-- 
GitLab

